[1단계 - HTTP 서버 구현하기] 엠제이(최민준) 미션 제출합니다

theminjunchoi opened the pull request:
반가워요 머랭~
리뷰 잘 부탁드려요☺️

저는 step1에서 객체지향보단, 기능 구현에 더 집중했던 것 같아요!
머랭이 이번 미션을 하면서 본인이 집중했던 부분, 신경썼던 부분에 대해 자유롭게 피드백 주시면 감사하겠습니다~!

미션1 화이팅이에요!

### 구현내용
- [x] GET /index.html 응답하기
- [x] CSS, JS 지원
- [x] Query String 파싱
- [x] 일치하는 계정, 비밀번호면 로그로 회원 조회 결과 출력

### 기타
- 테스트 코드에도 주석으로 적어놓았지만, "/"로 접근시에 자동으로 index.html을 띄워주기 위해 `Http11ProcessorTest.process`는 disabled 처리했습니다!
- POST 관련 기능은 Step2에 있어서 현재는 GET만 있다고 가정하고 작성했습니다!

theminjunchoi committed:
test: junit-jupiter-engine 의존성 비활성화

theminjunchoi committed:
test: FileTest에서 파일 경로 및 내용 읽기 테스트 추가

theminjunchoi committed:
test: IOStreamTest에서 InputStream 및 OutputStream 테스트 추가

theminjunchoi committed:
build: 테스트 의존성 정리 및 불필요한 라이브러리 제거

theminjunchoi committed:
feat: GET /index.html 응답하기 기능 추가

- Http11Processor에서 GET 요청 처리 로직 구현
- /index.html 요청 시 파일 내용을 응답으로 반환하는 readIndexFile 메서드 추가
- 요청이 유효하지 않을 경우 기본 응답으로 "Hello world!" 반환
- build.gradle에서 junit-jupiter 의존성 업데이트

theminjunchoi committed:
feat: css 및 정적 파일 요청 응답 기능 추가

- ContentTypeMapper 추가: 파일 확장자에 따른 Content-Type 매핑
- Http11Processor 수정: 정적 리소스(GET / 요청) 처리 로직 추가
   - ContentTypeMapper를 사용하여 Content-Type 동적으로 설정
   - 정적 리소스 파일 읽기용 readStaticFile 메서드 추가
- Http11ProcessorTest 수정: 테스트 파일 포맷팅 및 일부 테스트 비활성화

theminjunchoi committed:
feat: 서블릿 기반 요청 처리 및 ServletContainer 추가

- HttpRequest, HttpResponse 클래스 구현: HTTP 요청 및 응답 처리
- Servlet 인터페이스 및 구현체 추가: LoginServlet, StaticResourceServlet
   - LoginServlet: 로그인 요청 처리
   - StaticResourceServlet: 정적 리소스 제공
- ServletContainer 추가: URL 패턴 기반 서블릿 등록 및 요청 라우팅
- Http11Processor 수정: ServletContainer를 사용하여 요청 처리
- Connector 수정: 서블릿 등록 및 초기화 로직 추가
- 테스트 코드 수정: Servlet 기반 로직 테스트를 위해 StubSocket 및 ServletContainer 활용

theminjunchoi committed:
refactor: 조건문 간소화 및 코드 가독성 개선

theminjunchoi commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Override
+    public void service(final HttpRequest request, final HttpResponse response) {
+        if ("GET".equals(request.getMethod())) {
+            handleGet(request, response);
+            return;
+        }
+        
+        response.setStatus(405);
+        response.write("<html><body><h1>405 Method Not Allowed</h1></body></html>");
+    }
```
POST로 회원가입 하는 건 Step2에서 구현할게 토끼야

theminjunchoi commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,79 @@
+        response.write(loginHtml);
+    }
+
+    private void processLogin(final String account, final String password) {
+        final var userOptional = InMemoryUserRepository.findByAccount(account);
+
+        if (userOptional.isEmpty()) {
+            log.info("로그인 실패: 존재하지 않는 계정 - account: {}", account);
+            return;
+        }
+        
+        final var user = userOptional.get();
+        if (user.checkPassword(password)) {
+            log.info("로그인 성공: 회원 조회 결과 - {}", user);
+            return;
+        }
+        
+        log.info("로그인 실패: 비밀번호 불일치 - account: {}", account);
+    }
```
인정. 하지만 이건 톰캣 구현 미션이니까 무시할게
실제 서비스면 당연히 이렇게 안하지

theminjunchoi committed:
refactor: 정적 리소스 및 로그인 페이지 로드 로직 개선

- File 및 URL 기반 파일 읽기 방식을 InputStream으로 변경하여 리소스 로드 로직 단순화
- 에러 페이지 생성 로직 분리(createErrorPage 메서드 추가) 및 상태 코드 반영
- readLoginPage 및 readStaticFile 메서드에서 UTF-8 인코딩 적용
- 로그 메시지 추가: 리소스 파일 미존재 및 읽기 실패 시 경고 및 에러 로그 출력

theminjunchoi commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,79 @@
+            log.error("Failed to read login.html", e);
+            return "<html><body><h1>Error loading login page</h1></body></html>";
+        }
+    }
```
ㅇㅋ 반영해봄

theminjunchoi commented on tomcat/src/main/java/com/techcourse/servlet/StaticResourceServlet.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public void service(final HttpRequest request, final HttpResponse response) {
+        if (!"GET".equals(request.getMethod())) {
+            response.setStatus(405);
+            response.write("<html><body><h1>405 Method Not Allowed</h1></body></html>");
+            return;
+        }
+
+        String uri = request.getUri();
+
+        // 루트 경로면 index.html로
+        if ("/".equals(uri)) {
+            uri = "/index.html";
+        }
+
+        final String content = readStaticFile(uri);
+        final String contentType = ContentTypeMapper.get(uri);
+
+        response.setContentType(contentType);
+        response.write(content);
+    }
+
+    private String readStaticFile(final String uri) {
+        try {
+            final URL resource = getClass().getClassLoader().getResource("static" + uri);
+            if (resource == null) {
+                return "<html><body><h1>404 File Not Found</h1></body></html>";
+            }
+            return new String(Files.readAllBytes(new File(resource.getFile()).toPath()));
+        } catch (final Exception e) {
+            log.error("Failed to read static file: " + uri, e);
+            return "<html><body><h1>Error loading file</h1></body></html>";
+        }
+    }
+
+    @Override
+    public void destroy() {
+        log.info("StaticResourceServlet destroyed");
+    }
+}
```
음 이건 좀 더 고민해볼게

theminjunchoi commented on tomcat/src/main/java/com/techcourse/servlet/StaticResourceServlet.java:
```diff
@@ -0,0 +1,61 @@
+
+        response.setContentType(contentType);
+        response.write(content);
+    }
```
오 맞네 반영할게

theminjunchoi committed:
refactor: 정적 파일 읽기 및 응답 로직 개선

- StaticFileResult 레코드 추가: 파일 상태 코드, Content-Type, 내용 반환 구조 정리
- readStaticFile 메서드 수정: 에러 상태 코드 및 콘텐츠 타입 설정 로직 추가
- 코드 가독성 향상 및 예외 처리 개선: InputStream 사용 시 에러 핸들링 로직 강화
- StaticResourceServlet에서 응답 상태 코드 설정 로직 반영

theminjunchoi commented on tomcat/src/main/java/com/techcourse/servlet/StaticResourceServlet.java:
```diff
@@ -0,0 +1,61 @@
+            log.error("Failed to read static file: " + uri, e);
+            return "<html><body><h1>Error loading file</h1></body></html>";
+        }
+    }
```
ㅇㅋㅇㅋ 재밌다 반영해볼게

theminjunchoi committed:
feat: 경로 탐색 공격 방지 로직 추가

- readStaticFile 메서드에 경로 탐색 공격 방지 조건문 추가
  - ".." 및 "\\" 포함 여부를 확인하여 400 상태 코드 반환
  - 경고 로그 출력 및 에러 페이지 생성 로직 추가
- StaticResourceServlet 보안 강화 및 에러 처리 로직 개선

theminjunchoi commented on tomcat/src/main/java/org/apache/catalina/connector/Connector.java:
```diff
@@ -25,21 +28,31 @@ public Connector() {
+        container.addServlet("/*", new StaticResourceServlet());
+
+        return container;
+    }
```
이건 다음 미션하면서 고민해볼게~

theminjunchoi committed:
feat: HTTP 응답 상태 코드에 Reason Phrase 추가

- 상태 코드에 따른 Reason Phrase를 반환하는 getReasonPhrase 메서드 구현

theminjunchoi commented on tomcat/src/main/java/org/apache/catalina/ServletContainer.java:
```diff
@@ -0,0 +1,64 @@
+        return null;
+    }
+
+    private void handleNotFound(final HttpResponse response) {
+        response.setStatus(404);
+        response.write("<html><body><h1>404 Not Found</h1></body></html>");
+    }
```
ㅇㅋ 인정 반영해볼게

theminjunchoi commented on tomcat/src/main/java/org/apache/coyote/http11/HttpResponse.java:
```diff
@@ -0,0 +1,55 @@
+                    content);
+                    
+            outputStream.write(response.getBytes());
+            outputStream.flush();
```
확인함. 반영해볼게

theminjunchoi committed:
refactor: HttpResponse UTF-8 인코딩 적용

- StandardCharsets.UTF_8 사용하여 일관된 문자열 인코딩 적용
- 콘텐츠 길이 계산 및 응답 생성 시 UTF-8 기반 처리

cookie-meringue commented on study/build.gradle:
```diff
@@ -24,10 +24,6 @@ dependencies {
     implementation 'com.github.jknack:handlebars-springmvc:4.4.0'
 
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
-    testImplementation 'org.assertj:assertj-core:3.27.3'
-    testImplementation 'org.mockito:mockito-core:5.18.0'
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.13.3'
-    testImplementation 'org.junit.jupiter:junit-jupiter-engine:5.13.3'
```
이 의존성들을 제거한 이유는 무엇인가요?
엠제이가 겪었던 문제가 있다면 어떤 문제였고, 왜 의존성을 제거하는 선택을 하게 되었는지 궁금합니다!

theminjunchoi replied:
gradle버전과 Junit버전의 차이가 있어서 테스트가 실패했었는데요!
이를 해결하면서 보니 어차피 spring-boot-starter-test에서 테스트와 관련된 의존성을 같이 관리를 해주더라고요.
그래서 다 지워버렸습니다!

cookie-meringue commented on study/src/test/java/study/FileTest.java:
```diff
@@ -1,53 +1,52 @@
 
-        // todo
-        final String actual = "";
+        final URL resource = getClass().getResource("/" + fileName);
```
`"/"`를 추가하는 이유는 무엇인가요??

theminjunchoi replied:
클래스패스의 루트부터 찾기 위함입니다!

클래스패스 루트는 JVM이 클래스랑 리소스 파일 찾을 때 기준점이 되는 디렉터리를 말하는데요, `src/main/resources`와 `src/test/resources`가 각각 메인, 테스트의 클래스패스 루트가 됩니다.

"/"을 안해주면 현재 클래스와 같은 패키지 구조에서 찾게 됩니다. 즉, `src/test/resources/study/nextstep.txt` 이런 경로로 찾게 되는거죠.

하지만 저희가 원하는 파일의 위치는 `src/test/resources/nextstep.txt`이기  때문에 "/"를 붙여서 테스트의 클래스패스 루트에서 찾도록 해주었습니다.

cookie-meringue replied:
깔끔한 설명 좋습니다 👍

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Override
+    public void service(final HttpRequest request, final HttpResponse response) {
+        if ("GET".equals(request.getMethod())) {
+            handleGet(request, response);
+            return;
+        }
```
`HttpRequest`와 `HttpResponse` 를 추상화하신 것이 인상깊습니다~
이왕 하신 김에 Http Method 또한 추상화해보는 것은 어떨까요?

(3단계의 요구사항에 가까우니, 지금 단계에서는 넘어가도 좋습니다!)

theminjunchoi replied:
오 맞아요! 저도 HttpMethod도 객체로 분리를 하고 싶었는데, step2에서 POST를 구현하면서 같이 할 예정입니다.

cookie-meringue replied:
네 알겠습니다!

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private void processLogin(final String account, final String password) {
+        final var userOptional = InMemoryUserRepository.findByAccount(account);
+
+        if (userOptional.isEmpty()) {
+            log.info("로그인 실패: 존재하지 않는 계정 - account: {}", account);
+            return;
+        }
```
엠제이의 코드에서 `var` 키워드가 많이 보이는데요.
엠제이는 `var` 를 어떤 기준을 가지고 사용하시나요?
`Optional` 같은 타입에도 `var` 를 적용하시는 특별한 이유가 있다면 궁금합니다!

theminjunchoi replied:
사실 저는 타입을 명확히 해주면 좋다는 생각이라 이전까지는 잘 사용을 안했었는데요, 강의에서 구구가 사용하는 모습을 보고 직접 한 두곳에 사용을 해보면서 장단점을 느껴보고 싶었어요.

특히 저 부분은 `var`을 안쓰면 `Optional<User>`로 타입을 적어야하는데, 가독성을 위해 `var` 키워드를 사용했습니다. `Optional`을 사용해야하는 경우에는 `var`을 적는 것도 나쁘지 않은 것 같아요!

cookie-meringue replied:
넵 좋습니다!
지금까지 본 엠제이는 무언가를 사용할 때 대충 공부하고 넘어가지 않았을 것이라 생각됩니다.
`var` 에 대한 학습을 한 후, 엠제이만의 기준을 만들었을 것이라 생각됩니다!

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,92 @@
+        final var userOptional = InMemoryUserRepository.findByAccount(account);
+
+        if (userOptional.isEmpty()) {
+            log.info("로그인 실패: 존재하지 않는 계정 - account: {}", account);
+            return;
```
로그인 실패에 대한 처리는 다음 단계에서 구현하려고 로그만 남기는 것으로 이해했습니다!
제가 이해한 게 맞을까요?

theminjunchoi replied:
맞아요!

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -0,0 +1,92 @@
+        } catch (final IOException e) {
+            log.error("Failed to read login.html", e);
+            return createErrorPage("Error loading login page", 500);
+        }
```
`OOM` 발생에 대해 고려해보셨을까요?
`new String(inputStream.readAllBytes())` 방식은 코드가 간결하지만, `readAllBytes()`가 파일 전체를 Heap 메모리에 로드하므로 HTML 파일 사이즈가 크다면 `OOM` 이 발생할 수 있을 것 같습니다.
물론, 지금은 매우 간단한 서비스이고 로그인 HTML 파일의 사이즈는 `OOM` 을 유발할 만큼 크지 않겠지만, 생각해 볼 만한 문제인 것 같아 질문드립니다!

엠제이만의 생각이나 대안이 있다면 말씀해주시면 감사하겠습니다!

theminjunchoi replied:
아뇨! 머랭의 리뷰를 보고 좀 찾아봤는데요, 큰 파일이거나 동시사용자, 혹은 대용량의 이미지와 같은 경우 OOM이 발생할 수 있을 것 같아요. 물론 지금 환경에서는 그럴 문제가 없을 것 같지만, 스트리밍 방식으로 해결할 수 있을 것 같더라고요!

학습할 겸 한 번 구현해보겠습니다~!

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/Servlet.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface Servlet {
+    
+    void init();
+    
+    void service(HttpRequest request, HttpResponse response);
+    
+    void destroy();
```
`init()` 과 `destroy()` 의 존재 이유는 무엇인가요?
`jakarta` 의 `Servlet` 인터페이스와 거의 동일한 형태로 구현하신 것 같아 질문드립니다.

꼭 필요한 메서드인가요? 필요하다면 이유는 무엇인지 궁금합니다!

theminjunchoi replied:
실제 Servlet에서는 `init`과 `destory`에서 각각 DB 커넥션 풀 초기화와 정리, 설정 파일 로딩과 정리, 캐시 초기화와 정리, 외부 리소스 연결과 정리 등등 하는 게 많은 것 같더라고요!

당장 저 기능이 필요가 없어서 `init`과 `destory`를 안만들어도 되지만 구조를 참고할겸, 단순 로깅용으로 만들었습니다!

cookie-meringue replied:
넵 ㅎㅎ 그런 이유일 것 같아 질문드렀습니다!
저랑 관심 분야와 공부한 내용이 비슷해서 엠제이의 생각이 잘 읽히는 것 같아요 ㅎㅎ

엠제이도 경계하고 있겠지만, `Servlet` 과 같이 `"이미 잘 구현되어 있는 것"`을 참고하는 것이 엠제이만의 기준이나 스타일을 세우는 것에 방해가 되진 않을까 우려됩니다.

개인적으로 제가 이번 미션에서 경계하고 있는 부분이라 조심스럽게 말씀드립니다 ㅎㅎ

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/ServletContainer.java:
```diff
@@ -0,0 +1,64 @@
+
+    private static final Logger log = LoggerFactory.getLogger(ServletContainer.class);
+
+    private final Map<String, Servlet> servlets = new LinkedHashMap<>();
```
`LinkedHashMap` 을 선택하신 이유는 무엇인가요?

theminjunchoi replied:
`findServlet`에서 매칭을 해주고 있는데요,
우선적으로 정확한 매칭(`containsKey`)을 체크하고,
그 다음에 와일드카드 패턴을 체크하기 위해 순서보장을 해주었어요!

cookie-meringue replied:
효율적인 계산을 위해 `LinkedHashMap`을 사용하신거군요 ㅎㅎ

그런데, 순서 보장이 어떤 도움을 주는 것인지는 잘 모르겠습니다.
정확한 매칭인지 검사하는 시간복잡도 = O(1)
그게 아니라면, 와일드카드 기반으로 탐색하는 시간복잡도 = O(n)
`for`문에서 `다시 첫 번째 Entry부터 검사`를 하기 때문에 그렇게 느껴지는 것 같습니다!

제 생각이 엠제이와 다르다면 의견을 남겨주셔도 좋고, 중요한 부분은 아니라고 생각하신다면 넘어가셔도 좋을 것 같습니다! 이번 미션에서 중요한 부분은 아니니까요 ㅎㅎ

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/ServletContainer.java:
```diff
@@ -0,0 +1,64 @@
+
+    private final Map<String, Servlet> servlets = new LinkedHashMap<>();
+
+    public void addServlet(final String urlPattern, final Servlet servlet) {
```
URL 패턴을 기반으로 Servlet을 등록하신 이유는 무엇인가요?
Http Method 를 기준으로 등록할 순 없을까요?

theminjunchoi replied:
맞아요, 현재 구조에서는 예를 들어 /api/users로 요청이 오면 POST든 GET이든 같은 서블릿으로 가고 있어요.

물론 머랭이 말씀하신대로 Http method 기준으로도 매핑할 수 있겠지만 지금의 방식이 개인적으로 더 간단해보여서 지금처럼 설계했습니다.

머랭 리뷰를 보고 궁금해서 실제 톰캣 구조도 보았는데요, 이것도 URL 패턴기반으로 Servlet을 등록하는 것 같습니다.

cookie-meringue replied:
넵 ㅎㅎ 저도 `Servlet`(정확히는 `HttpServlet`) 이 URI 기반으로 등록되고 `doGET()`, `doPost()`... 등의 메서드로 Http Method에 따른 처리를 담당하는걸로 알고 있습니다!

엠제이가 이 패턴을 학습하시고 스스로의 기준을 가지고 적용하신건지, 만약 그렇다면 그 기준은 무엇인지 궁금했습니다!
엠제이의 의사결정 근거 이해했습니다 ㅎㅎ

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/ServletContainer.java:
```diff
@@ -0,0 +1,64 @@
+            return servlets.get(uri);
+        }
+
+        for (final Entry<String, Servlet> entry : servlets.entrySet()) {
+            final String pattern = entry.getKey();
+            if (pattern.endsWith("/*")) {
+                final String prefix = pattern.substring(0, pattern.length() - 2);
+                if (uri.startsWith(prefix)) {
+                    return entry.getValue();
+                }
+            }
+        }
```
"/**" 형식에 대한 처리도 고려하셨는지 궁금합니다!

theminjunchoi replied:
하하.. 안했습니다..ㅎㅎ
하다보니 점점 많아져서 처리해줄 게 많더라고요...!
배보다 배꼽이 커질 것 같아서 패스했어요

cookie-meringue replied:
ㅎㅎ 알겠습니다! 저도 그렇게 중요한 부분이라고 생각하지는 않아요!

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/ServletContainer.java:
```diff
@@ -0,0 +1,64 @@
+            }
+        }
+
+        return null;
```
`Optional` 사용에 대한 엠제이의 생각이 궁금합니다~ `null` 을 직접 다루도록 설게하신 이유가 있을까요??

theminjunchoi replied:
음 사실 별 차이가 없겠지만 성능을 고려했을 때 null 처리를 하는 게 더 좋아보였습니다.

spring mvc를 공부할 때 실제 `DispatcherServlet` 코드를 진짜 많이 봤었는데요, 이 코드에서도 Optinal 보다 null을 거의 대부분 사용하더라고요.

이런 라이브러리같은 경우에는 미세하게라도 성능을 중요시 생각하기 때문에 Optional 객체 생성 비용도 아끼면 좋지 않을까? 해서 null로 처리했어요!

cookie-meringue replied:
저도 `DispatcherServlet`과 더불어 Spring MVC와 관련된 코드를 재미있게 공부했습니다 ㅎㅎ.
비슷한 방향성을 가지고 공부한 크루를 만나서 재미있네요!

추가적으로, Spring 과 같은 프레임워크들은 하위호환성을 지키기 위해서도 `null` 을 사용할 것 같다는 생각도 듭니다!

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/connector/Connector.java:
```diff
@@ -25,21 +28,31 @@ public Connector() {
+        container.addServlet("/*", new StaticResourceServlet());
+
+        return container;
+    }
```
엠제이는 `Connector` 클래스의 역할이 무엇이라고 생각하나요??
`ServletContainer` 의 초기화와 더불어 `Servlet` 등록까지 `Connector`가 담당하도록 설계하신 이유가 궁금합니다!

> 지금 보니 코드래빗이 리뷰를 했던 부분이네요 ㅎㅎ
> 편하게 말씀해주시면 감사하겠습니다!

theminjunchoi replied:
지금보니 Connector의 역할이 많네요!

저는 Connector가 네트워크 연결만 처리하는 게 좋을 것 같아요.
서블릿 컨테이너의 초기화 부분만 담당하는 객체로 분리를 해볼게요!

cookie-meringue commented on tomcat/src/main/java/org/apache/coyote/http11/ContentTypeMapper.java:
```diff
@@ -0,0 +1,21 @@
+            return "application/octet-stream";
+        }
+        String ext = path.substring(dotIndex + 1);
+        return TYPES.getOrDefault(ext, "application/octet-stream");
```
기본 값 지정 👍👍
상수화해서 의미를 조금 더 명확하게 전달해도 좋을 것 같습니다!

theminjunchoi replied:
반영할게요!

cookie-meringue commented on tomcat/src/main/java/org/apache/coyote/http11/ContentTypeMapper.java:
```diff
@@ -0,0 +1,21 @@
+            "js", "application/javascript"
+    );
+
+    public static String get(String path) {
```
`path` 보다 더 의미를 잘 전달할 수 있는 매개변수명은 어떨까요??

파일 확장자를 검사하는 로직 같은데 매개변수명이 `path` 라서 경로 관련 로직인지 조금 헷갈렸습니다 😅

theminjunchoi replied:
앗 그러네요! `extension`이라는 이름으로 수정해볼게요!

cookie-meringue commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -27,21 +31,42 @@ public void run() {
+             final var outputStream = connection.getOutputStream();
+             final var reader = new BufferedReader(new InputStreamReader(inputStream))) {
+
+            final String requestLine = reader.readLine();
+            if (requestLine == null) {
+                return;
+            }
```
예외를 던지지 않고 `null` 을 반환하고 끝내는 이유는 무엇인가요??
아래 `parseRequest()` 메서드에서는 요청이 잘못된 경우 예외를 던지고 있길래 여쭤봅니다!

theminjunchoi replied:
클라이언트가 연결만 하고 아무것도 안보내거나, 연결이 갑자기 끊어지는 네트워크 에러와 같은 상황은 정상적인 상황으로 판단했어요! 실제 네트워크가 그렇게 안정적이지 않아서 null을 반환하는 경우는 당연히 생긴다고 생각했거든요! 

이와 다르게 parseRequest에서 던지는 에러는 요청 형식이 잘못된 것이기 때문에 에러라고 생각했습니다!

cookie-meringue replied:
넵 ㅎㅎ 좋습니다!

cookie-meringue commented on tomcat/src/main/java/org/apache/coyote/http11/HttpRequest.java:
```diff
@@ -0,0 +1,64 @@
+        this.parameters = parseParameters(queryString);
+    }
+    
+    private Map<String, String> parseParameters(final String queryString) {
+        final Map<String, String> params = new HashMap<>();
+        
+        if (queryString == null || queryString.isEmpty()) {
+            return params;
+        }
+        
+        final String[] pairs = queryString.split("&");
+        for (final String pair : pairs) {
+            final String[] keyValue = pair.split("=", 2);
+            if (keyValue.length == 2) {
+                try {
+                    final String key = URLDecoder.decode(keyValue[0], StandardCharsets.UTF_8);
+                    final String value = URLDecoder.decode(keyValue[1], StandardCharsets.UTF_8);
+                    params.put(key, value);
+                } catch (final Exception e) {
+                    // 파싱 실패시 무시
+                }
+            }
+        }
+        return params;
+    }
```
요청 파라메터 파싱 실패 시 무시하도록 구현하신 이유가 궁금합니다!

theminjunchoi replied:
어랏 원래는 로그 정도 남기려고 했는데 까먹었네요!
로그만 추가하겠습니다.

그리고 에러를 던지지 않은 이유는 다음과 같아요!
Http는 Best-Effort 프로토콜입니다. 브라우저나 클라이언트가 실수로 이상한 파라미터를 보내도 전체 요청을 날려먹으면 사용자 경험이 최악이 되기 때문에 이를 막아주고 싶었어요. 그래서 쿼리스트링 파싱 과정에서 일부 파라미터가 이상해도, 나머지는 정상처리될 수 있을 것 같아서 로그만 남기고 넘어가게 했습니다!

theminjunchoi replied:
또 이 리뷰를 보니 위에 적어주신 리뷰가 다시 생각이 들어서 보충설명할게요!

`parseParameters`과 달리 `parseRequest`는 에러를 던지고 있어요.
`parseRequest`는 HTTP 요청 라인을 파싱하는데, 이는 핵심 프로토콜 구조를 파싱하기 때문에 이게 잘못되면 HTTP 요청 자체가 의미가 없어서 에러를 던지게 설계했어요.

비유를 하자면 아래와 같은 의미일 것 같아요!
`parseRequest`: 편지 봉투 (받는 사람, 보내는 사람 주소) - 이게 틀리면 배달 불가
`parseParameters`: 편지 내용 일부 - 몇 글자 알아볼 수 없어도 나머지는 읽을 수 있음

cookie-meringue reviewed (changes requested):
고생 많으셨습니다 엠제이! 많이 배워갑니다!
스포?당하지 않기 위해 제 미션도 조금 끝내느랴 리뷰가 늦어졌네요 😅

객체지향보다는 기능 구현에 중점을 맞추셨다고 해서 객체지향적인 부분은 크게 리뷰를 남기지 않았습니다!

theminjunchoi committed:
refactor: BufferedReader를 사용하여 파일 읽기 개선

- InputStreamReader 및 BufferedReader 활용하여 리소스 파일 읽기 로직 변경
- lines()와 Collectors.joining()으로 파일 내용을 효율적으로 읽고 문자열 생성

theminjunchoi committed:
refactor: InputStream 사용 시 null 검증 로직 개선

- Objects.requireNonNull 사용하여 InputStream null 검증 로직 간소화
- 에러 처리 및 가독성 향상: 조건문 제거 및 코드 간결화
- LoginServlet과 StaticResourceServlet에 적용하여 일관성 확보

theminjunchoi committed:
refactor: 서블릿 컨테이너 초기화 로직 Context로 분리

- Context 클래스 추가: ServletContainer 초기화 및 서블릿 등록 로직 분리
- Connector 생성자 수정: Context를 통해 생성된 ServletContainer를 주입받도록 변경
- Tomcat 클래스 수정: Context를 사용하여 서블릿 컨테이너 초기화 및 종료 처리 로직 추가
- 코드 구조 개선 및 가독성 향상: 컨테이너 초기화 책임 분리로 유지보수성 강화

theminjunchoi committed:
feat: 파일 없으면 404 페이지 반환

theminjunchoi committed:
refactor: ContentTypeMapper 메서드 인자명 및 변수명 수정

theminjunchoi committed:
refactor: 파일 타입 기본값 상수화

theminjunchoi reviewed (commented):
안녕하세요 머랭!
리뷰 퀄리티가 너무 좋네요!

제가 생각만 하고 언급 안하고 넘어간 것까지 잘 캐치해주셔서 저도 다시 생각정리를 할 수 있었습니다!

theminjunchoi committed:
feat: HttpRequest에 로깅추가

cookie-meringue reviewed (approved):
step1 고생 많으셨습니다 엠제이!

스프링을 공부했던 방식과 공부 내용이 저와 비슷한 것이 느껴져서 재미있네요 ㅎㅎ
엠제이만의 이유와 기준을 알 수 있어 좋았습니다!

ps. 제 리뷰 텀이나 리뷰의 양, 질에 관해 조정이 필요하다고 생각하시면 편하게 DM 주세요!
